azurelinuxagent/ga/update.py

```diff
@@ -606,18 +519,18 @@ def _processing_new_extensions_goal_state(self):
         egs = self._goal_state.extensions_goal_state
         return self._goal_state is not None and egs.id != self._last_extensions_gs_id and not egs.is_outdated
 
-    def _process_goal_state(self, exthandlers_handler, remote_access_handler):
+    def _process_goal_state(self, exthandlers_handler, remote_access_handler, agent_update_handler):
         protocol = exthandlers_handler.protocol
 
         # update self._goal_state
         if not self._try_update_goal_state(protocol):
             # agent updates and status reporting should be done even when the goal state is not updated
-            self.__update_guest_agent(protocol)
-            self._report_status(exthandlers_handler)
+            agent_update_handler.run()
+            self._report_status(exthandlers_handler, agent_update_handler, gs_updated=False)
             return
 
-        # check for agent updates
-        self.__update_guest_agent(protocol)
+        gs_updated = self._processing_new_extensions_goal_state() or self._processing_new_incarnation()
+        agent_update_handler.run()
```

Comment:
If that's the case, I don't need protocol at all for agent update. All that can be retrieved from goal state if we pass it. This is going to effect on test cases and lot of test cases mocked the protocol and refreshed/updated goal state through protocol.

When we do refactor, we have to update few places like where we took protocol dependency instead of goal state like ext_handler and remote handler.

I'll happy to do that now for agent update but I'm thinking is it easy and utilize efforts very effectively if we do it together or knock it out one by one when we get a chance?